File: frontend/src/types/insights.ts
export type TrendDirection = "improving" | "declining" | "stable" | "needs_attention";

export interface ScoreEvolution {
  current: number;
  previous: number;
  best: number;
  average: number;
  improvement_pct: number | null;
}

export interface RiskAnalysis {
  current: string;
  previous: string;
  highest: string;
  lowest: string;
  trend: "improving" | "worsening" | "stable";
}

export interface TestTrend {
  direction: TrendDirection;
  latest_value: number;
  previous_value: number | null;
  data_points: number;
}

export interface InsightCard {
  type: string;
  icon: string;
  title: string;
  description: string;
}

export interface RecommendationsPayload {
  exercise: string[];
  nutrition: string[];
  hydration: string[];
  sleep: string[];
  follow_up_tests: string[];
  doctor_visit: string[];
}

export interface HistoryPoint {
  report_id: number;
  date: string;
  health_score: number;
  risk_level: string;
}

/** Discriminated union — every /insights* endpoint returns one of these two shapes. */
export type NotEnoughData = {
  has_enough_data: false;
  message: string;
};

export type FullInsightsResponse = {
  has_enough_data: true;
  reports_analyzed: number;
  score_evolution: ScoreEvolution;
  risk_analysis: RiskAnalysis | null;
  trends: Record<string, TestTrend>;
  summary: string;
  cards: InsightCard[];
};

export type TrendsResponse = {
  has_enough_data: true;
  trends: Record<string, TestTrend>;
};

export type RecommendationsResponse = {
  has_enough_data: true;
  recommendations: RecommendationsPayload;
};

export type HistoryResponse = {
  has_enough_data: true;
  history: HistoryPoint[];
};

File: backend/app/schemas/insights.py
from datetime import datetime
from typing import Optional, Literal

from pydantic import BaseModel


TrendDirection = Literal["improving", "declining", "stable", "needs_attention"]


class NotEnoughData(BaseModel):
    has_enough_data: Literal[False]
    message: str


class ScoreEvolution(BaseModel):
    current: int
    previous: int
    best: int
    average: float
    improvement_pct: Optional[float] = None


class RiskAnalysis(BaseModel):
    current: str
    previous: str
    highest: str
    lowest: str
    trend: Literal["improving", "worsening", "stable"]


class TestTrend(BaseModel):
    direction: TrendDirection
    latest_value: float
    previous_value: Optional[float] = None
    data_points: int


class InsightCard(BaseModel):
    type: str
    icon: str
    title: str
    description: str


class InsightsOverviewResponse(BaseModel):
    has_enough_data: Literal[True]
    reports_analyzed: int
    score_evolution: ScoreEvolution
    risk_analysis: Optional[RiskAnalysis] = None
    trends: dict[str, TestTrend]


class FullInsightsResponse(InsightsOverviewResponse):
    summary: str
    cards: list[InsightCard]


class TrendsResponse(BaseModel):
    has_enough_data: Literal[True]
    trends: dict[str, TestTrend]


class RecommendationsPayload(BaseModel):
    exercise: list[str]
    nutrition: list[str]
    hydration: list[str]
    sleep: list[str]
    follow_up_tests: list[str]
    doctor_visit: list[str]


class RecommendationsResponse(BaseModel):
    has_enough_data: Literal[True]
    recommendations: RecommendationsPayload


class HistoryPoint(BaseModel):
    report_id: int
    date: datetime
    health_score: int
    risk_level: str


class HistoryResponse(BaseModel):
    has_enough_data: Literal[True]
    history: list[HistoryPoint]